lib/puppet/type/tomcat_install.rb

```diff
@@ -0,0 +1,50 @@
+Puppet::Type.newtype(:tomcat_install) do
+  desc "Installs the software into the given directory."
+
+  ensurable
+
+  autorequire(:user) do
+    self[:user]
+  end
+
+  autorequire(:group) do
+    self[:group]
+  end
+
+  newparam(:catalina_home, namevar: true) do
+    desc "The directory of the Tomcat installation."
+
+    validate do |value|
+      raise ArgumentError, "Catalina home is required." if value.nil?
+    end
+  end
+
+  newparam(:version) do
+    desc "Specifies the owner of the source installation directory."
+
+    validate do |value|
+      raise ArgumentError, "Tomcat version is required." if value.nil? or value !~ /\d{1,2}\.\d{1,2}\.\d{1,2}/
+    end
+  end
+
+  newparam(:user) do
+    desc "Specifies the owner of the source installation directory."
+
+    defaultto 'tomcat'
+  end
+
+  newparam(:group) do
+    desc "Specifies the group of the source installation directory."
+
+    defaultto 'tomcat'
+  end
+
+  validate do
+    unless self[:ensure]
+      raise(Puppet::Error, "Ensure is a required attribute")
```

Comment:
this is not required as ensure will never be `nil`

Reply:
How do I default ensure? When I run this without ensure set it does not throw and error, but also doesn't do any work.

Reply:
sorry for the delay, here's an example of a default value for ensure https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/lib/puppet/type/file_line.rb#L91-L94